Proposal for unit & integration testing

robertdfrench opened the pull request:
There's a lot going on, and I haven't documented any of it. But the idea is to be able to do unit tests without relying on real git, since it could be slow, use the network, or impact local state.

So `cargo test --lib` will use mostly in-memory tests, reaching out to fake git binaries when necessary. The Integration tests (under the `tests/` folder) will run against the real git binary, and may do things like authenticate to remote hosts or create / remove branches.

This is not really intended to be a PR, so much as a conversation starter around testing. I think we'll want to do *some* of this, but maybe not all of it.

robertdfrench committed:
Draft: proposal for unit & integration testing

There's a lot going on, and I haven't documented any of it. But the idea
is to be able to do unit tests without relying on real git, since it
could be slow, use the network, or impact local state.

So `cargo test --lib` will use mostly in-memory tests, reaching out to
fake git binaries when necessary. The Integration tests (under the
`tests/` folder) will run against the real git binary, and may do things
like authenticate to remote hosts or create / remove branches.

calebwherry commented on failing_git.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
```
Relying on shell will be problematic cross-platform. We should re-write in Rust, these are simple programs so shouldn't add too much overhead.

calebwherry replied:
Fixed by #3.

calebwherry commented on src/lib.rs:
```diff
@@ -0,0 +1,100 @@
+    pub fn new() -> Git {
+        Git{ program: OsString::from("git") }
+    }
+    pub fn with_path(path: &Path) -> Git {
```
We should probably make this private so external users can't use it. It's really only for out unit test framework.

robertdfrench replied:
Fixed in 956844ef16b4f37e6c3cd4ddea27412609948f72 but I would like you to review.

calebwherry replied:
I like your solution. If we can move the impl to be test only, that works even better as separation to me.

calebwherry commented on src/lib.rs:
```diff
@@ -0,0 +1,100 @@
+    pub fn version(&self) -> io::Result<String> {
+        let output = Command::new(&self.program).arg("--version").output()?;
+        if !output.status.success() {
+            return Err(io::Error::new(io::ErrorKind::Other, "git --version"));
```
The actual error is lost here. Create a custom struct that includes exit status and anything else we need.

calebwherry replied:
Need to update all functions in this struct, not just this one.

robertdfrench replied:
This is fixed in 5f90c9f

calebwherry commented on src/lib.rs:
```diff
@@ -0,0 +1,100 @@
+
+pub fn extract_pr_names(branches: &str) -> Vec<String> {
+    lazy_static! {
+        static ref BEGINS_WITH_REMOTE_REF: Regex = Regex::new(r"^ *\** remotes/origin/").unwrap();
```
Using origin can cause issues with others not using that wordage. Replace with "not slash" match as it can be anything.

calebwherry replied:
Why not just drop the entire last bit and use `r"^ *\** remotes/"` as the pattern match? Does that potentially match branch where someone called it "remotes/"? Is that even possible?

robertdfrench replied:
Replying here [as well](https://github.com/robertdfrench/git-pr/issues/6#issuecomment-940565355) since I can't figure out this review interface: that regex is currently used for matching *and* trimming the the branch name. So if we changed the pattern to `r"^ *\** remotes/"`, then "origin" would end up as part of the PR's name:

```
* remotes/origin/pr-name/0  =>  origin/pr-name
```

BUT we could fix that by using [`String::split_once`](https://doc.rust-lang.org/std/string/struct.String.html#method.split_once) to split at the first "/" and just give us the 2nd half.

I think the [`\w`](https://docs.rs/regex/1.5.4/regex/#perl-character-classes-unicode-friendly) regex will get us part of the way there if we want to stick with a "pure regex" approach. It will match any character that can occur in a word, but not a hyphen, and remote names can include hyphens. So maybe something like `\w(-\w)?` could let us match remote names that contain hyphens.

calebwherry replied:
Yeah, that works for me.

robertdfrench replied:
Your original proposal was best: https://stackoverflow.com/questions/41461152/which-characters-are-illegal-within-a-git-remote-name/41462742

git doesn't explicitly say what characters can't be in a remote name, so the only thing we can really do is consume every character until the first '/', which is what c0a8dae6bd6ebaa6e261e9f97c1d18f97ecbcd66 does.

calebwherry commented on src/lib.rs:
```diff
@@ -0,0 +1,100 @@
+        static ref BEGINS_WITH_REMOTE_REF: Regex = Regex::new(r"^ *\** remotes/origin/").unwrap();
+        static ref ENDS_WITH_DIGIT: Regex = Regex::new(r"/\d+$").unwrap();
+    }
+    let pr_branches: Vec<&str> = branches.lines()
+        .filter(|b| BEGINS_WITH_REMOTE_REF.is_match(b))
+        .filter(|b| ENDS_WITH_DIGIT.is_match(b))
+        .collect();
+
+    let mut pr_names = vec![];
+    for branch in pr_branches {
+        let branch = BEGINS_WITH_REMOTE_REF.replace_all(&branch, "");
+        let branch = ENDS_WITH_DIGIT.replace_all(&branch, "");
+        pr_names.push(branch.to_string())
+    }
```
How do we combine these?

robertdfrench replied:
Following up with this in #7  - we discussed not resolving this issue for this PR, that still okay with you?

calebwherry replied:
Yep.

calebwherry commented on src/lib.rs:
```diff
@@ -0,0 +1,100 @@
+
+    #[test]
+    fn query_version_info() {
+        let path = Path::new(".").join("fake_git.sh");
```
Does just using the exe name work instead of having to join with the full path?

robertdfrench replied:
I'm resolving this one: the technical solution was introduced in e187ac1544f4b89a0830c3ba20b18f084d7b009c, but I've left #8 open as a style discussion in case we want to revisit.

calebwherry commented on src/lib.rs:
```diff
@@ -0,0 +1,100 @@
+        * stuff/I/wrote
+          trunk
+          remotes/origin/first-pr/0
+          remotes/origin/second/3
```
Add an additional test line for a "remote branch we are not actually tracking"

robertdfrench replied:
Fixed in 52c2d23 (#10)

calebwherry reviewed (changes requested):
Finished review.

Commit:
Fixes #3: Rewrite fake_git in Rust

* Rewrote fake_git sh scripts in Rust
* Moved all executables to src/bin & removed from TOML
* Updated tests to use new exes

robertdfrench committed:
Move `with_path` impl to tests module. Fixes #4.

I wasn't 100% sure this would work at first, but you can split an `impl`
block up over multiple sections. So since `with_path` doesn't need to be
defined for the `Git` struct in application code, I moved it into the
`test` module instead.

If you don't do something like this, then marking `with_path` as private
will generate a warning, since no other application code (even within
its own module) uses it.

Two alternative approaches:
* Leave `with_path` where it is, but mark it as `#[cfg(test)]`
* Change `new` to call `with_path`, so that `with_path` is used by something.

robertdfrench committed:
Test for remote refs which don't fit PR pattern.

robertdfrench commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+
+#[derive(Debug)]
+pub struct Git {
+    program: OsString,
```
Just change all of this to `String`.

calebwherry replied:
Fixes in 7e42b1f.

robertdfrench committed:
Add bare-minimum documentation

robertdfrench committed:
Custom error type for git invocations. Fixes #5.

Invoking git either raises an `io::Error` or returns a
`std::process::ExitStatus`, the `success()` method of which may return
`false`. The new type, `GitError`, is designed to encompass both of
these cases and should simplify error handling when interacting with git
subprocesses.

Commit:
Comment out Cargo.lock in gitignore. Fixes #12.

Commented Cargo.lock from gitignore so that we have a record
of why we did it. I think something like this is easy to forget
so leaving it along with the original comment makes sense.

Adding the version of Cargo.lock I get running on MacOs.

Commit:
Simplify pattern matching in fake_git. Fixes #13.

Commit:
Replace OsString/Path with String. Fixes #11.

calebwherry reviewed (commented):
All latest changes and comments resolved. Need final pass on open items.

robertdfrench committed:
Give rough overview in README

We can add build/release/platform info once we have CI set up (see #2).

robertdfrench committed:
Better example formatting. Fixes #14

robertdfrench commented:
@calebwherry I like all your changes, I think I addressed the remaining concerns you had. Only thing left is figuring out what path to take with matching remotes other than "origin".

robertdfrench commented:
Btw, here is the link to see what the updated README will look like: https://github.com/robertdfrench/git-pr/tree/pr-list/0

robertdfrench committed:
Match anything but '/' for origin name. Fixes #6.

As @calebwherry said in the original issue:

> Using origin can cause issues with others not using that wordage.
> Replace with "not slash" match as it can be anything.

I tried to be clever and match hyphenated words, but then I tried to
look up the actual definition of a remote name, and there does not seem
to be one: https://stackoverflow.com/questions/41461152/which-characters-are-illegal-within-a-git-remote-name/41462742

So, it seems like the best thing to do is detect whatever isn't a slash,
because we know those *can't* be allowed, since they are allowed in
branch names (and therefore any ref with slashes in the remote or
branch would parse ambiguously). Whatever else git tells us is obviously
valid, since git is the authority here.

robertdfrench committed:
Resolve a line-ordering conflict in README.md

calebwherry commented:
@robertdfrench Can you fix the merge conflicts? I assume just merge trunk is all that is needed? Once you do that, I'll merge this PR in!

robertdfrench commented:
@calebwherry good catch, and luckily this gives us a chance to see whether our new CI pipeline passes!

robertdfrench committed:
Test in release mode!

Commit:
Add more prints in GitHub YAML for debugging

Commit:
More debugging for PR failure

Commit:
Add setup repo step in GH actions yml. Fixes #17

GitHub checks out the repo in a detached head state. We use
a trick noted in the task to fix it.

Commit:
Add setup repo step in GH actions yml. Fixes #17

GitHub checks out the repo in a detached head state. We use
a trick noted in the task to fix it.

Try 2...

Commit:
Add setup repo step in GH actions yml. Fixes #17

GitHub checks out the repo in a detached head state. We use
a trick noted in the task to fix it.

Try 3...

Commit:
Print statment to help debug this more...

Commit:
Undo debugging for now...

robertdfrench committed:
Create working git repo for integration tests

robertdfrench committed:
Include tempdir crate

Found by @calebwherry.

robertdfrench committed:
Setup git config name and email

So we can commit!
